components/validator/challenger.go

```diff
@@ -219,6 +189,90 @@ func (c *Challenger) Stop() error {
 	return nil
 }
 
+func (c *Challenger) loop() {
+	defer c.wg.Done()
+
+	ticker := time.NewTicker(time.Minute)
+	defer ticker.Stop()
+
+	for ; ; <-ticker.C {
+		select {
+		case <-c.ctx.Done():
+			return
+		default:
+			shouldWait, err := c.shouldWaitSyncComplete()
+			if err != nil {
+				continue
+			}
+			if shouldWait {
+				continue
+			}
+			c.log.Info("challenge: kroma node sync completed")
+
+			if c.cfg.ChallengerEnabled {
+				if err := c.updateCheckpoint(); err != nil {
+					continue
+				}
+			}
+
+			if err := c.scanPrevOutputs(); err != nil {
+				c.log.Error("challenge: failed to scan previous outputs", "err", err)
+				continue
+			}
+
+			// if challenge mode on, subscribe L2 output submission events
+			if c.cfg.ChallengerEnabled {
+				c.wg.Add(1)
+				go c.subscribeL2OutputSubmitted()
+			}
+
+			// subscribe challenge creation events
+			c.wg.Add(1)
+			go c.subscribeChallengeCreated()
+
+			return
+		}
+	}
+}
+
+// updateCheckpoint updates checkpoint which is the last checked output index, so the next output handling starts after
+// this point. If checkpoint is behind the latest output index, handle the previous outputs from the checkpoint.
+func (c *Challenger) updateCheckpoint() error {
+	cCtx, cCancel := context.WithTimeout(c.ctx, c.cfg.NetworkTimeout)
+	defer cCancel()
+	nextOutputIndex, err := c.l2ooContract.NextOutputIndex(utils.NewSimpleCallOpts(cCtx))
+	if err != nil {
+		c.log.Error("challenger: failed to get the latest output index", "err", err)
+		return err
+	}
+	if nextOutputIndex.Cmp(common.Big0) == 0 {
+		// if no outputs have been submitted, set checkpoint to 1 because genesis output cannot be challenged
+		c.checkpoint = common.Big1
+	} else {
+		// set checkpoint to latestOutputIndex (nextOutputIndex - 1)
+		c.checkpoint = new(big.Int).Sub(nextOutputIndex, common.Big1)
+	}
+	c.metr.RecordChallengeCheckpoint(c.checkpoint)
+	return nil
+}
+
+func (c *Challenger) shouldWaitSyncComplete() (bool, error) {
```

Comment:
Wouldn't it be nice to have the outputSubmitter, challenger, and guardian all apply shouldWaitSyncComplete initially. aa Wouldn't it be nice to have shouldWaitSyncComplete initially in common with the NewValidator?

Reply:
Good idea. I'll check the sync status initially in `validator.go`, and remove this code blocks from each component.